lib/common/socket.c

```diff
@@ -934,7 +951,21 @@ static void proceed_handshake(h2o_socket_t *sock, const char *err)
     on_handshake_complete(sock, err);
 }
 
-void h2o_socket_ssl_handshake(h2o_socket_t *sock, SSL_CTX *ssl_ctx, const char *server_name, h2o_socket_cb handshake_cb)
+h2o_iovec_t get_ssl_session_cache_key(h2o_socket_t *sock, const char *server_name)
+{
+    struct sockaddr sa;
+    h2o_socket_getpeername(sock, &sa);
+    uint16_t port = ntohs(((struct sockaddr_in *)&sa)->sin_port);
```

Comment:
There is no guarantee that `h2o_socket_getpeername` would succeed, nor that the returned type is `sockaddr_in`.

I believe you should for IPv4 and IPv6 addresses extract the port numbers (in appropriate ways), while refusing to use session resumption in all other cases.

And considering these requirements you _might_ want to reconsider how the generation of the key is extracted to a dedicated function.